models/packages/package_version.go

```diff
@@ -37,24 +38,30 @@ type PackageVersion struct {
 
 // GetOrInsertVersion inserts a version. If the same version exist already ErrDuplicatePackageVersion is returned
 func GetOrInsertVersion(ctx context.Context, pv *PackageVersion) (*PackageVersion, error) {
-	e := db.GetEngine(ctx)
+	for i := 0; i <= 5; i++ {
+		inserted, err := db.InsertOnConflictDoNothing(ctx, pv)
+		if err != nil || inserted {
+			return pv, err
+		}
 
-	key := &PackageVersion{
-		PackageID:    pv.PackageID,
-		LowerVersion: pv.LowerVersion,
-	}
+		key := &PackageVersion{
+			PackageID:    pv.PackageID,
+			LowerVersion: pv.LowerVersion,
+		}
 
-	has, err := e.Get(key)
-	if err != nil {
-		return nil, err
-	}
-	if has {
-		return key, ErrDuplicatePackageVersion
-	}
-	if _, err = e.Insert(pv); err != nil {
-		return nil, err
+		has, err := db.GetEngine(ctx).Get(key)
+		if has {
+			return key, ErrDuplicatePackageVersion
+		} else if err != nil {
+			return key, err
+		}
+		// This really should never happen and can only happen if this function
+		// is being called outside of a transaction and between the on conflict insert failing
+		// the conlicting item is removed.
+		//
```

Comment:
There are many `for i := 0; i <= 5; i++` , I do not think it's necessary.

"between the insert failing and conflicting item is removed.": Is there really a case for it? Failing fast without retry is also the correct behavior for all cases.

Reply:
And I believe the retry would case more problems:

Think about 2 threads, they do the same thing:

1. for loop try to insert
2. do something, and insert some related data
2.1 if fail, remove the inserted row
2.2 then, remove other rows related to the inserted row
3. if succeed, over

When thread-B is on step2.1, thread-A is on step1. then thread-A will continue to do something and insert related data

Then thread-A comes to step2, and thread-B comes to 2.2. 

Then thread-B would delete thread-A's data incorrectly.

Reply:
There is no remove.

Reply:
Then why retry? Why would it ever succeed?

Reply:
> I am simply suggesting how it is that this case could possibly happen - otherwise its impossible

I do not think it would ever occur. If it's impossible, the dead code should be removed.

Reply:
It would succeed in the described case.

Imagine the function was called outside a transaction with goroutine A, and a separate goroutine B was also running.

1. Goroutine A tries to INSERT ON CONFLICT but finds that the requested object is in the DB
2. Goroutine B then deletes it.
3. Goroutine A would then try to SELECT * but then finds that the requested object is no longer in the db

What would you prefer to happen then? Would you prefer that we fail? Or would you prefer that we retry?

I wrote retry because I thought reviewers would prefer that we retried. However, if you'd prefer that we simply failed it's very simple to change it.

Reply:
The question is why `Goroutine B then deletes it.` ? 

And even there is a  retrying for 5 times, the problem still happens, Goroutein B could delete the data after Goroutine A's 5-time-retrying. The problem just becomes more complex, no benefit to developers.


It's caller's duty to make sure the data insertion/deletion logic are complete.